types/index.d.ts

```diff
@@ -2084,13 +2104,30 @@ declare module 'mongoose' {
   type TreatAsPrimitives = actualPrimitives |
   NativeDate | RegExp | symbol | Error | BigInt | Types.ObjectId;
 
+  // This will -- when possible -- extract the original type of the subdocument in question
+  type LeanSubdocument<T> = T extends (Types.Subdocument<Require_id<T>['_id']> & infer U) ? LeanDocument<U> : Omit<LeanDocument<T>, '$isSingleNested' | 'ownerDocument' | 'parent'>;
+
   type LeanType<T> =
     0 extends (1 & T) ? T : // any
       T extends TreatAsPrimitives ? T : // primitives
-        T extends Types.Subdocument ? Omit<LeanDocument<T>, '$isSingleNested' | 'ownerDocument' | 'parent'> : // subdocs
+        T extends Types.Subdocument ? LeanSubdocument<T> : // subdocs
           LeanDocument<T>; // Documents and everything else
 
-  type LeanArray<T extends unknown[]> = T extends unknown[][] ? LeanArray<T[number]>[] : LeanType<T[number]>[];
+  // Used only when collapsing lean arrays for ts performance reasons:
+  type LeanTypeOrArray<T> = T extends unknown[] ? LeanArray<T> : LeanType<T>;
```

Comment:
Does this still work with https://github.com/Automattic/mongoose/issues/10293 ? We need to be careful about nested arrays.

Reply:
Pretty sure it does, but I will verify. I did see the need for nested arrays and I tested it, but I'll look at the specific issue and make sure.

Reply:
I have added a test which both verifies that #10293 is still fixed and also to ensure that it doesn't happen again in the future.